towhee/dataframe/iterators.py

```diff
@@ -0,0 +1,146 @@
+import weakref
+import time
+
+from towhee.dataframe.dataframe_v2 import DataFrame
+
+
+class MapIterator:
+    """
+    A row-based `DataFrame` iterator.
+    """
+    def __init__(self, df: DataFrame, block = False):
+        self._df_ref = weakref.ref(df)
+        self._offset = 0
+        self._block = block
+        self._id = df.register_iter(self)
+        self._done = False
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns:
+            (`Tuple[Any, ...]`)
+                In the normal case, the iterator will return a `Tuple` at each call.
+            (`None`)
+                In the case that the `DataFrame` is not sealed and the new rows are
+                not ready yet, the iterator will return `None`. The caller should
+                determine whether to block the iteration or exit the loop.
+        Raises:
+            (`StopIteration`)
+                The iteration end iff the `DataFrame` is sealed and the last row is
+                reached.
+        """
+        if self._done:
+            raise StopIteration
+
+        df = self._df_ref()
+
+        code, row = df.get(self._offset, count = 1, iter_id = self._id)
+
+        if code == 'Index_GC':
+            raise IndexError
+
+        elif code == 'Index_OOB_Unsealed':
+            if self._block:
+                while code == 'Index_OOB_Unsealed':
```

Comment:
We couldn't use while-loops and sleep to wait for data. If one image-pipeline has 10 ops, the worst case we may sleep 10s, but one image maybe only need 100ms.

Reply:
And if a big pipeline has hundreds of iters, the pipeline-process will do nothing but while-loops.

Reply:
Is there a good tactic for this other than condition vars? Condition vars wont transfer well to the distributed version. For the waiting I only set it to 1s for testing purposes. It can be much lower, just needs a balance between checking to much and waiting too long. 

Furthermore, we will probably want to keep this logic within the iterator and not within dataframe.

Reply:
We can't use while-loops if we case about latency.